silentsoul884/BetterEvents.js

```diff
@@ -0,0 +1,175 @@
+                            eventName: {
+                                type: "string"
+                            },
+                            data: {
+                                type: "string",
+                                defaultValue: "{}"
+                            }
+                        }
+                    },
+                    {
+                        blockType: Scratch.BlockType.HAT,
+                        id: _BetterEvent.id,
+                        opcode: "on",
+                        text: "When i receive event [eventName]",
+                        arguments: {
+                            eventName: {
+                                type: Scratch.ArgumentType.STRING
+                            }
+                        },
+                        isEdgeActivated: false
+                    },
+                    {
+                        blockType: Scratch.BlockType.REPORTER,
+                        opcode: "data",
+                        text: "Get Event Data [eventName]",
+                        arguments: {
+                            eventName: {
+                                type: Scratch.ArgumentType.STRING
+                            }
+                        }
+                    }
+                ]
+            };
+        }
+        emit({ eventName, data }, util) {
+            this.eventData.set(eventName, data);
+            this.events.emit(eventName, data);
+            this.events.on(eventName, () => {
+                util.startHatsWithParams(`${_BetterEvent.id}_on`, {
+                    parameters: { eventName }
+                });
+            });
+        }
+        on({ eventName }, util) {
+            const { hatParam } = util.thread;
+            if (hatParam.eventName === eventName) {
+                return eventName;
+            }
+            return null;
+        }
+        data({ eventName }) {
+            return this.eventData.get(eventName);
+        }
+    };
+
+    // src/index.ts
+    (function(Scratch2) {
+        if (Scratch2.extensions.unsandboxed === false) {
+            throw new Error("Sandboxed mode is not supported");
+        }
+        if (Scratch2.vm?.runtime) {
+            Scratch2.extensions.register(new BetterEvent(Scratch2.runtime));
+        } else {
+            window.tempExt = {
+                Extension: BetterEvent,
+                info: {
+                    extensionId: BetterEvent.id,
+                    name: BetterEvent.id,
+                    description: BetterEvent.description,
+                    featured: true,
+                    disabled: false,
+                    collaboratorList: []
+                },
+                l10n: {
+                    en: {
+                        [BetterEvent.id]: BetterEvent.description,
+                        [BetterEvent.description]: BetterEvent.description
+                    }
+                }
+            };
+        }
+    })(Scratch);
```

Comment:
Gandi-IDE has opened vm to all extensions, this method used to differentiate between Turbowarp and Gandi-IDE is not working. This method of distinguishing between Turbowarp and Gandi-IDE does not work, and may result in missing registration information in Gandi.